### ISSUE TITLE
Steam always reprocesses shader precache at game launch

### ISSUE THREAD

**Issue by felipec** [ValveSoftware/steam-for-linux#9748]
This is the same as #7306, which was apparently wrongly closed and locked.

I have noticed it happening in the past few weeks. I don't know if it was happening before.

#### Your system information

* Steam client version: 1.0.0.78 (1687386907)
* Distribution:  Arch Linux
* Opted into Steam client beta?: No
* Have you checked for system updates?: Yes
* Steam Logs: [steam-logs.tar.gz](https://github.com/ValveSoftware/steam-for-linux/files/11898486/steam-logs.tar.gz)
* GPU: NVIDIA GeForce RTX 3060 Laptop

#### Please describe your issue in as much detail as possible:

Every time I launch Dota 2 it says "Processing Vulkan Shaders" and spends several minutes there. Not every time there's an update, every time I launch it.

Why does it need to do that every single time?

#### Steps for reproducing this issue:

1. Launch Dota 2

**Comment by kisak-valve** [comment]
Hello @felipec, the general topic of shader pre-cache updates is being tracked at #8076, not #7306. Noted at https://github.com/ValveSoftware/steam-for-linux/issues/8076#issuecomment-1430641159, shader pre-cache updates are designed to be small incremental updates as shaders are seen in the wild.

The majority of the time, most of the shaders have already been seen by the video driver and stored in the video driver specific on-disk shader cache.

I took a look at this a bit and did not reproduce the issue locally with Steam running on an Intel GPU. I would expect when the `Steam` -> `Settings` -> `Downloads` -> `Allow background processing of Vulkan shaders` option is enabled, that would allow the Steam client to work ahead and not need to handle that at game start, and when that option is disabled, for it to run as needed.

Another NVIDIA user repeated the same basic conditions, and did reproduce the issue of processing shader on every game launch, which hints that this is a shader precache subcomponent + NVIDIA specific issue. Can you note what video driver version you're using?

**Comment by apocalyptech** [comment]
FWIW, I was the other NVIDIA user confirming the conditions.  My details:

OpenGL renderer string: NVIDIA GeForce GTX 1060 6GB/PCIe/SSE2
OpenGL core profile version string: 4.6.0 NVIDIA 535.54.03
OpenGL core profile shading language version string: 4.60 NVIDIA

**Comment by 3331** [comment]
This issue appears to have been fixed for me with July 10th update, however some games such as CS:GO still appear to be recompiling a bunch of shader during launch, since I have to wait 3-4 minutes before the game is ready to play. This used to only happen once when nvidia drivers was updated/changed, but now it happens every time CS:GO is launched.

**Comment by JU12000** [comment]
Can confirm this issue is also happening to me even on the July 10 update.

Steam client version: 1.0.0.78-2 (1689034492)
Distribution: Manjaro Linux
Opted into Steam client beta?: No
Have you checked for system updates?: Yes
OpenGL renderer string: NVIDIA GeForce GTX 1080/PCIe/SSE2
OpenGL core profile version string: 4.6.0 NVIDIA 535.54.03
OpenGL core profile shading language version string: 4.60 NVIDIA

Tried with multiple different games in immediate succession and each launch kicks off the Vulkan Shader processor even with `Allow background processing of Vulkan shaders` turned on.

**Comment by JU12000** [comment]
Don't want to spam but some more (possibly) relevant info came up.

After rebooting Manjaro, Steam would no longer launch for me. A quick search and compare in `journalctl` told me the difference was:
`Could not connect to X session manager: Authentication Rejected, reason: None of the authentication protocols specified are supported and h`

Searching for that lead me to [this Steam Support page](https://help.steampowered.com/en/faqs/view/2112-4548-B98B-3A06) which informed me to run Steam with `DISPLAY=:0 steam`.

Fortunately this worked, at which point I turned off `Allow background processing of Vulkan shaders` and attempted to exit and start steam normally, which also worked.

I checked `echo $DISPLAY` before running the `DISPLAY=:0 steam` command and after launching Steam the second time, it was and still is set globally to `:0.0`.

I'm not sure how closely related the two are but thought it might help to share.

**Comment by adamnejm** [comment]
I experience the same thing.  
Additionally these Counter-Strike: Global Offensive issues might be related:  
- After an update, the games takes 1-2 minutes to start, but now it happens every time the shader processing pops up.  
- Seemingly at random you may load into the game and have 15 FPS, which restarting the game does not fix, but restarting Steam client does.  
  
**System:**
OS: ArchLinux
Kernel: 6.4.3-arch1-1
Steam: 1.0.0.78-1 from official *multilib* repositories
Steam build: 1689034492 (July 10th 2023 @ 23:59:09)

**Launch options:**
`gamemoderun %command% -novid -nojoy -freq 144 -tickrate 128`

Related GPU variables:
```sh
export NVD_BACKEND="direct"                                                      # VA-API backend
export LIBVA_DRIVER_NAME="nvidia"                                                # VA-API driver
export PROTON_HIDE_NVIDIA_GPU=0                                                  # Report as Nvidia
export PROTON_ENABLE_NVAPI=1                                                     # Enable NVAPI library
export PROTON_ENABLE_NGX_UPDATER=1                                               # Enable OTA updates for Nvidia NGX
export VKD3D_CONFIG="dxr,dxr11"                                                  # Enable DXR and DXR 1.1
```

**Comment by Zulleyy3** [comment]
I think I am experiencing the same issue with Apex Legends for a long time. The process itself usually takes multiple hours to run anywhere close to completion, and gets stuck around 63% (which is where i often just stop it because it's been multiple hours). Not sure if this is relevant, but skipping compilation and playing in the game's training range for a couple minutes makes the game perfectly playable without any issues whatsoever.

I am currently checking monitoring my shader_log and shader_log_previous, the compilation also seems to get stuck at 61%-63%
Is this normal?
[shader_log.previous.txt](https://github.com/ValveSoftware/steam-for-linux/files/12064747/shader_log.previous.txt)

(the attached files are from a compilation process that is currently running. I will see if I can update this later once it's hopefully done)
---- 
From my shader_log_previous: 
[2023-07-13 15:35:19] Still replaying 1172470 (61%, 3145066/3697507).
[...]
[shader_log.txt](https://github.com/ValveSoftware/steam-for-linux/files/12064745/shader_log.txt)

[2023-07-13 16:00:40] Still replaying 1172470 (61%, 3146346/3697507).

In my current shader log (and the rollover):
[2023-07-16 19:49:10] Still replaying 1172470 (63%, 4116215/4623168).
[...]
[2023-07-16 20:12:28] Still replaying 1172470 (63%, 4866506/5334698).
[2023-07-16 20:12:28] Still replaying 1172470 (63%, 4866506/5334698).
[2023-07-16 20:12:38] Still replaying 1172470 (63%, 4873024/5338397).
[2023-07-16 20:12:48] Still replaying 1172470 (63%, 4877809/5340212).
[2023-07-16 20:12:58] Still replaying 1172470 (63%, 4882259/5349464).
[2023-07-16 20:13:08] Still replaying 1172470 (63%, 4888509/5356871).

System:
OS: ArchLinux
Kernel: 6.4.3-arch1-1
CPU: Ryzen 2700X 
Nvidia: RTX 3070 on 535.54.03-8 
Steam: 1.0.0.78-1 from official multilib repositories
Steam build:  1689034492 (July 16th 2023 @ 19:69 CEST)

**Comment by 3331** [comment]
> This issue appears to have been fixed for me with July 10th update, however some games such as CS:GO still appear to be recompiling a bunch of shader during launch, since I have to wait 3-4 minutes before the game is ready to play. This used to only happen once when nvidia drivers was updated/changed, but now it happens every time CS:GO is launched.

What I said here was incorrect, the issue reappears once I resume from suspended system or reboot the machine.

**Comment by jaxwilko** [comment]
Is there any update on this? I've been having the same issue for a while, every time I restart my machine and try and launch any game it starts `Processing Vulkan shaders`. Even if I let it run to completion, after the machine is restarted it will try and rebuild the shaders again.

**System:**
```
OS: Debian GNU/Linux trixie/sid x86_64 
Kernel: 6.4.0-2-amd64
DE: GNOME 43.6 
WM: Mutter 
CPU: Intel i9-9900K (16) @ 5.000GHz 
GPU: NVIDIA GeForce RTX 2080 Ti Rev. A 
```

**Comment by kisak-valve** [comment]
Hello @jaxwilko, what NVIDIA driver version are you using?

It looks like this specific issue was a video driver bug regarding how the driver handled its own on-disk shader cache. It's possible that the issue tracked here was introduced with 535.43.02 and quietly resolved in 535.86.05. I don't know off hand if other release series were affected.

**Comment by adamnejm** [comment]
Not the previous guy, but I face the same issue with `nvidia` package, ver. `535.98` from official Arch Linux repositories along with `6.4.10-arch1-1` kernel. Previously I've also used `535.86.05` and it was affected as well.

**Comment by jaxwilko** [comment]
@kisak-valve thanks for replying, even on a sunday, absolute legend <3.

Sorry for not incluiding this in my initial comment, my current nvidia driver version is: `525.125.06`.

I'm 99% sure I've had the same issue with the following versions, but cannot be 100% sure:
- `525.105.17`
- `525.89.02`
- `525.85.12`

For Debian testing `525.125.06` is the latest available driver, see: https://packages.debian.org/testing/nvidia-driver

Is there anything else I can try or provide that would help?

**Comment by adamnejm** [comment]
@kisak-valve So sorry for the confusion and wasting your time, but turns out I was wrong earlier.  
  
I just tested 4 games: **Counter-Strike: Global Offensive**, **Left 4 Dead 2**, **Deep Rock Galactic** and **FTL: Faster Than Light** using `nvidia-535.98-2` and normal, non-beta Steam release, ver. `1690583737`.  
The shader cache seems to be working perfectly, even after restarting Steam or the whole PC.  
  
I think the *Need Retest* label should be re-added so other people can confirm this.

**Comment by felipec** [comment]
> Can you note what video driver version you're using?

I suppose in June 28 I was using `nvidia 535.54.03`.

Back then I was **certainly** experiencing some issue. At some point I don't know when the amount times I saw the processing message was significantly reduced, to the point that I don't consider it an issue any more.

Now I'm using `nvidia 535.98` and and `steam 1690583737`.

I just updated Dota 2, launched it, and didn't even show the processing message.

If memory serves I still see the message, but it's certainly not **every** time, it's only sometimes, and it doesn't take forever.

I'm pretty sure there was an issue, but whatever it was I don't think it's present anymore, at least on my system.

**Comment by adamnejm** [comment]
Since my last update, 3/4 games - **CS:GO** (twice), **L4D2** (once), **DRG** (once), **FTL** (none) had popped up with the shader precaching, although in each case it didn't take very long *(~30s avg on 5600X)*, so I assume it wasn't compiling them from scratch, that's good.  
  
When it comes to **DRG** and **L4D2** they had updates pushed during that time, however **CS:GO** didn't.  
Assuming shader precaching may be triggered after a game update, it could be theoretically possible that there was an internal build of **CS2 Limited Test** which caused shader precaching *(my account participates in the **CS2** beta)*, just a speculation though.  
  
As @felipec stated, overall, the current state is acceptable.

**Comment by AllDragonsAreDead** [comment]
Adding `__GL_SHADER_DISK_CACHE_SKIP_CLEANUP=1` to ~/.profile helped me solve the issue. Before, I had to spend an hour compiling shaders after every Dota restart. Now Dota starts quickly, as expected.
Source: https://www.reddit.com/r/linux_gaming/comments/9gly07/nvidia_shadercache_use_this_env_variable_to_not/

Apparently, NVIDIA driver cleans/removes shaders specific to an app, when their size exceeds 128 MB. The solution mentioned above is one of the workarounds, but probably not the best one. Some people mention that game developers actually can fix that on their end, so users don't need to change driver's settings. Sadly, I can't find the link explaining that now.

**Comment by TLATER** [comment]
> Apparently, NVIDIA driver cleans/removes shaders specific to an app, when their size exceeds 128 MB.

This is a lifesaver. Of course it's NVIDIA's fault, it always is. I don't think I'll ever buy another one of their devices.

I've set up that env variable for the games that were constantly rebuilding their shaders, and banished them to specific directories with `__GL_SHADER_DISK_CACHE_PATH`. No more rebuilds setting my CPU to 100% for 6+ hours, no more framerate stuttering whenever I enter uncached regions. Thank you so much.

**Comment by Balssh** [comment]
Thanks! `__GL_SHADER_DISK_CACHE_SKIP_CLEANUP=1` solved it for me.

**Comment by g572staem** [comment]
>  Steam -> Settings -> Downloads -> Allow background processing of Vulkan shaders 

Jeepers, talk about illogical placement of an option.

**Comment by kisak-valve** [comment]
Closing as the issue tracked here was an NVIDIA sided disk cache regression around 535.54.03 which has been resolved somewhere around the NVIDIA 535.98 driver update.
